flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializerSnapshotData.java

```diff
@@ -0,0 +1,288 @@
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+
+import static org.apache.flink.util.LinkedOptionalMap.optionalMapOf;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * This class holds the snapshot content for the {@link PojoSerializer}.
+ *
+ * <h2>Serialization Format</hr>
+ *
+ * <p>The serialization format defined by this class is as follows:
+ *
+ * <pre>{@code
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |                                            POJO class name                                          |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |      number of fields      |                (field name, field serializer snapshot)                 |
+ * |                            |                                pairs                                   |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |         number of          |       (registered subclass name, subclass serializer snapshot)         |
+ * |   registered subclasses    |                                pairs                                   |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |         number of          |     (non-registered subclass name, subclass serializer snapshot)       |
+ * | non-registered subclasses  |                                pairs                                   |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * }</pre>
+ */
+@Internal
+final class PojoSerializerSnapshotData<T> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(PojoSerializerSnapshotData.class);
+
+	// ---------------------------------------------------------------------------------------------
+	//  Factory methods
+	// ---------------------------------------------------------------------------------------------
+
+	/**
+	 * Creates a {@link PojoSerializerSnapshotData} from configuration of a {@link PojoSerializer}.
+	 *
+	 * <p>This factory method is meant to be used in regular write paths, i.e. when taking a snapshot
+	 * of the {@link PojoSerializer}. All POJO fields, registered subclass classes, and non-registered
+	 * subclass classes are all present.
+	 */
+	static <T> PojoSerializerSnapshotData<T> createFrom(
+			Class<T> pojoClass,
+			LinkedHashMap<Field, TypeSerializer<?>> fieldSerializers,
+			LinkedHashMap<Class<?>, TypeSerializer<?>> registeredSubclassSerializers,
+			HashMap<Class<?>, TypeSerializer<?>> nonRegisteredSubclassSerializers) {
+
+		LinkedHashMap<Field, TypeSerializerSnapshot<?>> fieldSerializerSnapshots = new LinkedHashMap<>(fieldSerializers.size());
+		fieldSerializers.forEach((k, v) -> fieldSerializerSnapshots.put(k, v.snapshotConfiguration()));
```

Comment:
Should we use here TypeSerializerUtils#snapshotBackwardsCompatible ?

Reply:
good catch! Yes, we should be using `TypeSerializerUtils#snapshotBackwardsCompatible`